src/Microsoft.ML.Core/Data/ReadOnlyMemoryUtils.cs

```diff
@@ -0,0 +1,464 @@
+﻿using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Internal.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Microsoft.ML.Runtime.Data
+{
+    public static class ReadOnlyMemoryUtils
+    {
+
+        /// <summary>
+        /// This method retrieves the raw buffer information. The only characters that should be
+        /// referenced in the returned string are those between the returned min and lim indices.
+        /// If this is an NA value, the min will be zero and the lim will be -1. For either an
+        /// empty or NA value, the returned string may be null.
+        /// </summary>
+        public static string GetRawUnderlyingBufferInfo(out int ichMin, out int ichLim, ReadOnlyMemory<char> memory)
+        {
+            MemoryMarshal.TryGetString(memory, out string outerBuffer, out ichMin, out int length);
+            ichLim = ichMin + length;
+            return outerBuffer;
+        }
+
+        public static int GetHashCode(this ReadOnlyMemory<char> memory) => (int)Hash(42, memory);
+
+        public static bool Equals(this ReadOnlyMemory<char> memory, object obj)
+        {
+            if (obj is ReadOnlyMemory<char>)
+                return Equals((ReadOnlyMemory<char>)obj, memory);
+            return false;
+        }
+
+        /// <summary>
+        /// This implements IEquatable's Equals method. Returns true if both are NA.
+        /// For NA propagating equality comparison, use the == operator.
+        /// </summary>
+        public static bool Equals(ReadOnlyMemory<char> b, ReadOnlyMemory<char> memory)
+        {
+            if (memory.Length != b.Length)
+                return false;
+            Contracts.Assert(memory.IsEmpty == b.IsEmpty);
+
+            MemoryMarshal.TryGetString(memory, out string outerBuffer, out int ichMin, out int length);
+            int ichLim = ichMin + length;
+
+            MemoryMarshal.TryGetString(b, out string bOuterBuffer, out int bIchMin, out int bLength);
+            int bIchLim = bIchMin + bLength;
+            for (int i = 0; i < memory.Length; i++)
+            {
+                if (outerBuffer[ichMin + i] != bOuterBuffer[bIchMin + i])
+                    return false;
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Does not propagate NA values. Returns true if both are NA (same as a.Equals(b)).
```

Comment:
"Returns true if both are NA" suggests this is copy-pasted from somewhere, since of course now we will no longer have the notion of an NA string. This suggests an incomplete conversion, and is an opportunity for code simplification that we should do right now.